src/pages/SearchPage/SearchPage.test.js

```diff
@@ -0,0 +1,210 @@
+    const { getByTestId } = setup();
+    const button = getByTestId('search-button');
+    const input = getByTestId('search-input');
+    expect(input.value).toEqual(DEFAULT_SUBREDDIT);
+    expect(button).toBeDisabled();
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+});
+
+describe('Error and heatmap display states', () => {
+  test('Will show an error message when there are no results', async () => {
+    const { getByTestId, findByText } = setup();
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    await findByText('0 results returned.');
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+  test('Will show an error message when the API call fails', async () => {
+    const { getByTestId, findByText } = setup('fail');
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    await findByText('There was an error processing your request.');
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+  test('Will load heatmap with real test data', async () => {
+    const { getByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    expect(getByTestId('heatmap')).toBeInTheDocument();
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+});
+
+describe('Heatmap values', () => {
+  test('Heatmap values matches dummy data', async () => {
+    const dummyHeatmapValues = getHeatmapValues(dummyPosts);
+    const { getByTestId, getAllByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    // Checking buttons
+    const heatmapValues = getAllByTestId('heatmap-button').map((button) => parseInt(button.textContent, 10));
+    // Make sure number of buttons in heatmap is correct
+    expect(heatmapValues.length).toEqual(NUM_GRID_ITEMS);
+    // Heatmap values must match dummy data values
+    expect(heatmapValues).toEqual(dummyHeatmapValues);
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+  test('Heatmap elements use the correct theme colours', async () => {
+    const { getByTestId, getAllByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    // Checking buttons
+    const heatmapButtons = getAllByTestId('heatmap-button');
+    for (let i = 0; i < heatmapButtons.length; i += 1) {
+      const button = heatmapButtons[i];
+      const buttonValue = parseInt(button.textContent, 10);
+      if (buttonValue <= 10) {
+        expect(button).toHaveStyleRule('background', theme.heatmapColors[buttonValue]);
+      } else {
+        expect(button).toHaveStyleRule('background', theme.heatmapColors[10]);
+      }
+    }
+  });
+  test('Heatmap displays the correct timezone message', async () => {
+    const { getByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
```

Comment:
Since this assertion is repeated often: a nice way to prevent this kind of repetition is to create a function inside setup. 

```
const setup = () => {
  const queries = render(...)

  const waitForData = async() => {
    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
  }
	
  return {
    ...queries,
    waitForData,
  }
}
```

and then use it here like

```
await waitForData();
```

This way it's a bit more descriptive what the assertion is doing.

You can also create functions that are more complex like `searchNewSubreddit` where you enter a new subreddit and  submit the form or so. Just as an example, not saying that you should do that here.

Here is a nice blog post [AHA testing](https://kentcdodds.com/blog/aha-testing)